### PR TITLE
table cell action when value is a button

### LINKED ISSUE #945: Clickable Links in a Table
**What would that feature address**
It is not possible to show clickable links in a table

***Description of the ideal solution***
It would be great to be able to tell Taipy that a column's data type is "link" in the table definition to make http links clickable. And it would be nice to pass markdown or html links into the table to have a link text and url property available.

***Caveats***
Maybe passing markdown or html links into the table is too much. Setting the data type to "link" would be great already!

***Other options***
-

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Create related issue in taipy-doc for documentation and Release Notes
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it
- [ ] Ensure any change is well documented

### LINKED ISSUE #947: BUG-Empty value is valid despite restricted enumeration in table cell
**Description**

Since Taipy 3.1, table cell values can be restricted.
However, when you edit a table cell, you can use the Clear button (next to the Open button accessing the drop-down menu with predefined values), or simply select all the cell content and press de Delete key.
The cell is then set to the empty string.

Now if an empty string is *not* in the column's lov (and that lov does not have the None value), it is accepted by the table.

**Expected behavior**

The empty cell, if "" is not in the column's lov, should not be accepted.

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Create related issue in taipy-doc for documentation and Release Notes if relevant

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
- a button is shown when cell value is \[button label](button value)
- fix lov clear selection
resolves #945
resolves #947
```

import pandas as pd

from taipy.gui import Gui, State

data = {
    "label": ["red", "white"],
    "timedelta": [pd.Timedelta(days=1), pd.Timedelta(days=2)],
    "links": ["[link 1](a value)", "[another link](http://www.google.com)"],
}

col_lov = ["value 1", "value 2", "value 3"]

def on_action(state: State, var_name: str, payload: dict):
    print(f"on_action(state, var_name: {var_name}, payload: {payload})")


def on_edit(state: State, var_name: str, payload: dict):
    print(f"on_edit(state, var_name: {var_name}, payload: {payload})")

md = """
<|{data}|table|lov[label]={col_lov}|on_action=on_action|on_edit=on_edit|>
"""

Gui(md).run()

```

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/tableUtils.tsx [hunk 1]
```diff
@@ -291,55 +286,75 @@ export const EditableCell = (props: EditableCellProps) => {
...
-                if (val === null) {
-                    castedVal = val;
-                } else if (isValid(val)) {
-                    castedVal = dateToString(getTimeZonedDate(val as Date, formatConfig.timeZone, withTime), withTime);
-                } else {
-                    return;
-                }
-                break;
+    const button = useMemo(() => {
+        if (onSelection && typeof value == "string" && value.startsWith("[") && value.endsWith(")")) {
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/tableUtils.tsx:299]: I would have used a regexp here...
But well, line 301 makes sure we're good, kind of...

File: frontend/taipy-gui/src/components/Taipy/tableUtils.tsx [hunk 2]
```diff
@@ -291,55 +286,75 @@ export const EditableCell = (props: EditableCellProps) => {
...
-            );
-        setEdit((e) => !e);
-    }, [onValidation, val, rowIndex, colDesc.dfid, colDesc.type, formatConfig.timeZone, withTime]);
+        return undefined;
+    }, [value, onSelection]);
+
+    const onCheckClick = useCallback(
+        (evt?: MouseEvent<HTMLElement>) => {
+            evt && evt.stopPropagation();
+            let castedVal = val;
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/tableUtils.tsx]:
```suggestion
            let castVal = val; // Please
```

File: taipy/gui/viselements.json [hunk 3]
```diff
@@ -1111,7 +1111,7 @@
...
+            "doc": "The name of a function that is triggered when the user selects a row.<br/>All parameters of that function are optional:\n<ul>\n<li>state (<code>State^</code>): the state instance.</li>\n<li>var_name (str): the name of the tabular data variable.</li>\n<li>payload (dict): the details on this callback's invocation.<br/>This dictionary has the following keys:\n<ul>\n<li>action: the name of the action that triggered this callback.</li>\n<li>index (int): the row index.</li>\n<li>col (str): the column name.</li>\n<li>reason (str): the origin of the action: cmick or button.</li>\n<li>value (str): the value associated with the button when reason == button and original value is [label](button value).</li></ul></li></ul>.",
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]:
```suggestion
            "doc": "The name of a function that is triggered when the user selects a row.<br/>All parameters of that function are optional:\n<ul>\n<li>state (<code>State^</code>): the state instance.</li>\n<li>var_name (str): the name of the tabular data variable.</li>\n<li>payload (dict): the details on this callback's invocation.<br/>This dictionary has the following keys:\n<ul>\n<li>action: the name of the action that triggered this callback.</li>\n<li>index (int): the row index.</li>\n<li>col (str): the column name.</li>\n<li>reason (str): the origin of the action: "click", or "button" if the cell contains a Markdown link syntax.</li>\n<li>value (str): the *link value* indicated in the cell when using a Markdown link syntax (that is, <i>reason</i> is set to "button").</li></ul></li></ul>.",
```

What is 'payload.value' when *reason* is "click"?

**Review by FabienLelaquais — DISMISSED** [review]
I like that!

File: frontend/taipy-gui/src/components/Taipy/tableUtils.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/tableUtils.tsx:299]: simple is beautiful

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**